### PR TITLE
Actor critic algo

### PR DESCRIPTION
**Pull request by lgraesser**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -81,7 +81,7 @@ For auto-syncing `data/` files, we use Gulp. This sets up a watcher for automati
 ```shell
 npm install --global gulp-cli
 npm install --save-dev gulp gulp-watch gulp-changed
-run the
+# run the file watcher
```
Comment (kengz) [inline README.md]: For this file, take upstream instead of HEAD since I just updated the entire README and fixed them

File: circle.yml [hunk 2]
```diff
@@ -17,7 +17,7 @@ dependencies:
     - pip install pandas
   override:
     - pip install $TF_BINARY_URL
-    - pip install git+https://github.com/tflearn/tflearn.git
+    - pip install git+https://github.com/Theano/Theano.git
```
Comment (kengz) [inline circle.yml]: same here, take upstream over HEAD

File: requirements.txt [hunk 3]
```diff
@@ -4,7 +4,7 @@ scipy==0.18.0
 matplotlib==1.4.3
 pandas==0.18.1
 https://storage.googleapis.com/tensorflow/mac/cpu/tensorflow-0.11.0rc1-py3-none-any.whl
-git+https://github.com/tflearn/tflearn.git
+git+https://github.com/Theano/Theano.git
```
Comment (kengz) [inline requirements.txt]: also here

**Comment by kengz** [comment]
oh dont merge this yet, until the keras imports are contained within classes.

File: rl/agent/actor_critic.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        error = np.broadcast_to(error, (error.shape[0], self.env_spec['action_dim']))
+        # Mask for actions not taken
+        error = np.multiply(error, minibatch['actions'])
+        loss_critic = self.critic.model.train_on_batch(minibatch['states'], curr_targets)
+        loss_actor = self.actor.model.train_on_batch(minibatch['states'], error)
+        return loss_critic + loss_actor
+
+    def save(self, model_path, global_step=None):
+        logger.info('Saving model checkpoint')
+        self.critic.model.save_weights(os.path.join(model_path, "_critic"))
```
Comment (kengz) [inline rl/agent/actor_critic.py]: beware of memleaks, keras doesn't seem to handle files IO reliably. Is there a way to pass around model weights without saving external files? would fix dqn_freeze too

**Comment by kengz** [comment]
also might be related, this solved the continuous control, right? is this an implementation of [the ddpg paper](https://arxiv.org/abs/1509.02971)?

File: rl/agent/actor_critic.py (hunk 4 above)
Comment (lgraesser) [inline rl/agent/actor_critic.py]: Makes sense - I'll look into this. Also am not going to merge this soon, just wanted to update the branch as a lot had changed in master.

**Comment by lgraesser** [comment]
Yes, definitely related. The algorithms are essentially the same. The ddpg uses a few extra tricks - lagged target network (DQN freeze) and batch normalization. Also we would need to change the way the NNs are parameterized to account for continuous action spaces.

**Comment by kengz** [comment]
ooook just read the paper for real; the rough plan:
1. get DPG to work, + policy gradient
2. extend DPG like how DDPG extends it, by adding batch normalization and separate actor/critic target networks.

**Comment by lgraesser** [comment]
I think we should delete this. It should be fairly easy to implement once DDPG works and the structure of DDPG is much better. What do you think?

**Comment by kengz** [comment]
agreed. go ahead
